bob-common/src/metrics/mod.rs

```diff
@@ -228,20 +229,31 @@ fn install_global_recorder(recorders: Vec<Box<dyn Recorder>>) {
 }
 
 #[allow(unused)]
-fn install_prometheus() {
-    let recorder = build_prometheus();
+async fn install_prometheus() {
+    let recorder = build_prometheus().await;
     metrics::set_boxed_recorder(Box::new(recorder)).expect("Can't set Prometheus recorder");
 }
 
-fn build_prometheus() -> PrometheusRecorder {
-    metrics_exporter_prometheus::PrometheusBuilder::new()
+async fn build_prometheus() -> PrometheusRecorder {
+    let (recorder, exporter) = metrics_exporter_prometheus::PrometheusBuilder::new()
         .listen_address(
             "0.0.0.0:9000"
                 .parse::<SocketAddr>()
                 .expect("Bad metrics address"),
         )
         .idle_timeout(MetricKindMask::ALL, Some(Duration::from_secs(2)))
-        .build()
+        .build_with_exporter()
+        .expect("Failed to set Prometheus exporter");
+
+    tokio::spawn(async move {
+        pin!(exporter);
+        loop {
+            select! {
+                _ = &mut exporter => {}
+            }
+        }
+    });
```

Comment:
This looks a little bit strange. Why is this needed?

Reply:
I just checked [source code of metrics-prometheus-exporter builder's `install` method](https://github.com/metrics-rs/metrics/blob/main/metrics-exporter-prometheus/src/builder.rs#L202) and found it there. 

As I understood (hope that I understood right), this code is needed because exporter has to be polled to process requests from prometheus server, which it [promethes server] sends to scrape metrics. Without this code incoming 'scraping' requests won't be processed

Reply:
As I see, a thread with busy loop started here, and that thread do nothing (`{}`). I don't understand, where the 'scraping' requests processing performed. 

@pyakushin maybe you can understand how this block of code works

Reply:
[in tokio docs](https://docs.rs/tokio/1.11.0/tokio/macro.select.html) you may find this pattern for select:
`<pattern> = <async expression> (, if <precondition>)? => <handler>,`
and in our case `<async expression>` is handler itself (*), so if it completed (**), that means, that request is handled, and `{}` in this case is some kind of postprocessing (which we don't need to do)

(*) - this logic can be found [here](https://github.com/metrics-rs/metrics/blob/main/metrics-exporter-prometheus/src/builder.rs#L276). You see, that exporter itself - future (or 'async block'), which prepares some stuff and calls `server.serve(..)` at the end.

(**) You may read about this [here](https://rust-lang.github.io/async-book/06_multiple_futures/03_select.html#interaction-with-unpin-and-fusedfuture). That's not `select!` which is used here, because I use the one from tokio, but logic is the same.

Reply:
@ikopylov

Reply:
Ok, now I understand how it works. Thanks!